File: framework/security/src/main/java/top/bettercode/simpleframework/security/config/ApiSecurityConfiguration.java
package top.bettercode.simpleframework.security.config;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import javax.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer;
import org.springframework.context.MessageSource;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder;
import org.springframework.security.access.SecurityConfig;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import top.bettercode.simpleframework.security.ApiToken;
import top.bettercode.simpleframework.security.ApiSecurityErrorHandler;
import top.bettercode.simpleframework.security.ApiTokenService;
import top.bettercode.simpleframework.security.DefaultAuthority;
import top.bettercode.simpleframework.security.IResourceService;
import top.bettercode.simpleframework.security.repository.ApiTokenRepository;
import top.bettercode.simpleframework.security.repository.InMemoryApiTokenRepository;

@Configuration(proxyBeanMethods = false)
@ConditionalOnWebApplication
public class ApiSecurityConfiguration {

  private final ApiSecurityProperties securityProperties;

  public ApiSecurityConfiguration(
      ApiSecurityProperties securityProperties) {
    this.securityProperties = securityProperties;
    DefaultAuthority.DEFAULT_AUTHENTICATED = new SecurityConfig(
        securityProperties.getDefaultAuthority());
    DefaultAuthority.DEFAULT_GRANTED_AUTHORITY = new SimpleGrantedAuthority(
        securityProperties.getDefaultAuthority());
  }

  @ConditionalOnMissingBean(IResourceService.class)
  @Bean
  public IResourceService resourceService() {
    return new IResourceService() {
    };
  }

  @ConditionalOnMissingBean(PasswordEncoder.class)
  @Bean
  public PasswordEncoder passwordEncoder() {
    return new BCryptPasswordEncoder();
  }

  @Bean
  public ApiTokenService apiTokenService(ApiTokenRepository apiAuthorizationService,
      UserDetailsService userDetailsService) {
    return new ApiTokenService(securityProperties, apiAuthorizationService, userDetailsService);
  }

  @Bean
  public ApiSecurityErrorHandler securityOAuth2ErrorHandler(MessageSource messageSource,
      @Autowired(required = false) HttpServletRequest request) {
    return new ApiSecurityErrorHandler(messageSource, request);
  }

  @ConditionalOnMissingBean(ApiTokenRepository.class)
  @Bean
  public ApiTokenRepository apiAuthorizationService() {
    Cache<String, ApiToken> cache = Caffeine.newBuilder()
        .expireAfterWrite(Math.max(securityProperties.getAccessTokenValiditySeconds(),
            securityProperties.getRefreshTokenValiditySeconds()), TimeUnit.SECONDS)
        .maximumSize(10000).build();
    Cache<String, String> accessTokenBuild = Caffeine.newBuilder()
        .expireAfterWrite(securityProperties.getAccessTokenValiditySeconds(), TimeUnit.SECONDS)
        .maximumSize(10000).build();
    Cache<String, String> refreshTokenBuild = Caffeine.newBuilder()
        .expireAfterWrite(
            securityProperties.getRefreshTokenValiditySeconds(), TimeUnit.SECONDS)
        .maximumSize(10000).build();
    return new InMemoryApiTokenRepository(cache.asMap(), accessTokenBuild.asMap(),
        refreshTokenBuild.asMap());
  }

  @Configuration(proxyBeanMethods = false)
  @ConditionalOnWebApplication
  protected static class ObjectMapperBuilderCustomizer implements
      Jackson2ObjectMapperBuilderCustomizer {

    @Override
    public void customize(Jackson2ObjectMapperBuilder jacksonObjectMapperBuilder) {
      jacksonObjectMapperBuilder.serializerByType(GrantedAuthority.class,
          new JsonSerializer<GrantedAuthority>() {
            @Override
            public void serialize(GrantedAuthority value, JsonGenerator gen,
                SerializerProvider serializers)
                throws IOException {
              gen.writeString(value.getAuthority());
            }
          });
    }
  }

}

File: security/src/main/kotlin/top/bettercode/summer/security/config/ApiSecurityConfiguration.kt
package top.bettercode.summer.security.config

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider
import com.github.benmanes.caffeine.cache.Caffeine
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication
import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.MessageSource
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import top.bettercode.summer.security.ApiTokenService
import top.bettercode.summer.security.IResourceService
import top.bettercode.summer.security.client.ClientDetailsService
import top.bettercode.summer.security.repository.InMemoryStoreTokenRepository
import top.bettercode.summer.security.repository.StoreTokenRepository
import top.bettercode.summer.security.support.ApiSecurityErrorHandler
import top.bettercode.summer.security.token.AccessTokenConverter
import top.bettercode.summer.security.token.DefaulAccessTokenConverter
import top.bettercode.summer.security.token.StoreToken
import java.security.SecureRandom
import java.util.concurrent.TimeUnit
import javax.servlet.http.HttpServletRequest
import kotlin.math.max

@Configuration(proxyBeanMethods = false)
@ConditionalOnWebApplication
@EnableConfigurationProperties(ApiSecurityProperties::class)
class ApiSecurityConfiguration(
        private val securityProperties: ApiSecurityProperties
) {
    @ConditionalOnMissingBean(IResourceService::class)
    @Bean
    fun resourceService(): IResourceService {
        return object : IResourceService {}
    }

    @ConditionalOnMissingBean(AccessTokenConverter::class)
    @Bean
    fun accessTokenConverter(): AccessTokenConverter {
        return DefaulAccessTokenConverter()
    }

    @ConditionalOnMissingBean(ClientDetailsService::class)
    @Bean
    fun clientDetailsService(): ClientDetailsService {
        return ClientDetailsService(listOf(securityProperties))
    }

    @ConditionalOnMissingBean(PasswordEncoder::class)
    @Bean
    fun passwordEncoder(): PasswordEncoder {
        val secureRandomSeed = securityProperties.secureRandomSeed
        if (!secureRandomSeed.isNullOrBlank()) {
            return BCryptPasswordEncoder(-1, SecureRandom(secureRandomSeed.toByteArray()))
        }
        return BCryptPasswordEncoder()
    }

    @Bean
    fun apiTokenService(
            storeTokenRepository: StoreTokenRepository,
            clientDetailsService: ClientDetailsService,
            accessTokenConverter: AccessTokenConverter,
            userDetailsService: UserDetailsService
    ): ApiTokenService {
        return ApiTokenService(storeTokenRepository, clientDetailsService, accessTokenConverter, userDetailsService)
    }

    @Bean
    fun apiSecurityErrorHandler(
            messageSource: MessageSource,
            @Autowired(required = false) request: HttpServletRequest?
    ): ApiSecurityErrorHandler {
        return ApiSecurityErrorHandler(messageSource, request)
    }

    @ConditionalOnMissingBean(StoreTokenRepository::class)
    @Bean
    fun storeTokenRepository(clientDetailsService: ClientDetailsService): StoreTokenRepository {
        val cache = Caffeine.newBuilder()
                .expireAfterWrite(max(clientDetailsService.maxAccessTokenValiditySeconds,
                        clientDetailsService.maxRefreshTokenValiditySeconds).toLong(), TimeUnit.SECONDS)
                .maximumSize(10000).build<String, StoreToken>()
        val accessTokenBuild = Caffeine.newBuilder()
                .expireAfterWrite(clientDetailsService.maxAccessTokenValiditySeconds.toLong(), TimeUnit.SECONDS)
                .maximumSize(10000).build<String, String>()
        val refreshTokenBuild = Caffeine.newBuilder()
                .expireAfterWrite(
                        clientDetailsService.maxRefreshTokenValiditySeconds.toLong(), TimeUnit.SECONDS)
                .maximumSize(10000).build<String, String>()
        return InMemoryStoreTokenRepository(cache.asMap(), accessTokenBuild.asMap(),
                refreshTokenBuild.asMap())
    }

    @Configuration(proxyBeanMethods = false)
    @ConditionalOnWebApplication
    protected class ObjectMapperBuilderCustomizer : Jackson2ObjectMapperBuilderCustomizer {
        override fun customize(jacksonObjectMapperBuilder: Jackson2ObjectMapperBuilder) {
            jacksonObjectMapperBuilder.serializerByType(GrantedAuthority::class.java,
                    object : JsonSerializer<GrantedAuthority>() {
                        override fun serialize(
                                value: GrantedAuthority, gen: JsonGenerator,
                                serializers: SerializerProvider
                        ) {
                            gen.writeString(value.authority)
                        }
                    })
        }
    }
}